feat(2794): account permission selection

EtherWizard33 opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
This PR adds the ability to edit account permissions for in app browser dapps.



<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
4. What is the improvement/solution?
-->

## **Related issues**

Contributes to solve issue: https://github.com/MetaMask/MetaMask-planning/issues/2796

## **Manual testing steps**

I have added screenshot below to see how it looks.

1. Go to the in-app brower on the mobile wallet, connect to a dapp like uniswap
2. once connected click to top right icon to open the bottom sheet and start editing
3. examples in screenshots below

## **Screenshots/Recordings**


| Add 1 account       | Remove 1 account    |
|--------------|--------------|
| <img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/user-attachments/assets/21982551-8332-4596-89ed-50443ed86898"> |<img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/user-attachments/assets/e6bd540b-bafa-4883-a80a-bc138fcabcd1"> |



| Connect all      | Disconnect all   |
|--------------|--------------|
| <img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/user-attachments/assets/f2eaa78d-671c-4ea9-af0e-05a5a548631c"> |<img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/user-attachments/assets/14c43919-bbc0-49d3-beb9-62452c65b899"> |



| Add many     | Remove many   |
|--------------|--------------|
| <img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/user-attachments/assets/e0452e2a-8bb8-4741-8ec8-3642e63baeab"> |<img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/user-attachments/assets/7639f3c2-c214-4b25-8e38-8b4c243efba5"> |


## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

EtherWizard33 committed:
feat: on first load of the accounts checkbox list, preselect the permitted accounts

EtherWizard33 committed:
feat: add the ability to disconnect one and many

EtherWizard33 committed:
style: lint

tommasini commented on app/components/Views/AccountPermissions/AccountPermissions.tsx:
```diff
@@ -398,12 +447,20 @@ const AccountPermissions = (props: AccountPermissionsProps) => {
-    () => (
+  const renderEditAccountsPermissionsScreen = useCallback(() => {
+    let effectiveSelectedAddresses;
+    if (isFirstRender.current) {
```
Sorry I'm with low context here, why are we showing different addressed based on the render number could you elaborate?

EtherWizard33 replied:
Hey there! 😊

Yes, these lines are meant to initialize the checkbox list with all user accounts, with the permitted ones checked.

This is the part I'd like to try moving into a `useEffect` hook instead of using the `isFirstRender` ref.

To answer your question: `selectedAccounts` is the state variable updated by the checkbox list child component—it works well. However, before the checkbox list renders for the first time, `selectedAccounts` is empty. That's why I initialize the list with the permitted addresses checked initially. On subsequent renders, when users check or uncheck items, I use the `selectedAccounts` state variable, and the existing check/uncheck mechanism works as expected. 👍

tommasini replied:
Do you want to try to move it on this PR or another on the future? I see, it's like a trick for you to switch from the context of the selected address coming from redux to the interaction on the checkbox of permissions. That is for me, although using a useEffect and create a variable with a name more intuitive with the addesses checked it would be better. I will not block the PR will approve it now

EtherWizard33 committed:
Merge branch 'main' into feat

EtherWizard33 committed:
Merge branch 'main' into feat

EtherWizard33 committed:
style: lint

EtherWizard33 committed:
chore: temparily excluding test converage on two files until feature complete, we plan to start writing tests next week

EtherWizard33 committed:
Merge branch 'main' into feat

EtherWizard33 committed:
feat: putting behind feture flag the new account removal code meant for multichain

EtherWizard33 committed:
fix: only happy paths of selection deselection were working, this is now fixed

And was clearing the selected accounts of the edit all accounts flow, and removing accounts was buggy if all permitted were removed before adding a new permitted account, all this is fixed in this commit

EtherWizard33 committed:
style: remove dead code

EtherWizard33 committed:
fix: disabled button wasn't enabled after user clicked select all, this is now fixed

EtherWizard33 committed:
feat: add the ability to hide the disconnect all button for example when this component is used in the context of connecting more accounts

EtherWizard33 committed:
fix: when no accounts are in the list, hide the select all and the description telling the user to select an account when there is none to select

EtherWizard33 committed:
style: lint

EtherWizard33 committed:
merge main into feat

EtherWizard33 committed:
style: lint

EtherWizard33 committed:
feat: put behind feature flag the disabling of button when no changes has occured, this feature is only for when feature flag is on

EtherWizard33 committed:
feat: ensure all new screens used from the permission screen are behind feature flag
